Import for PLE/PLP

anthonysena opened the pull request:
Aims to add import methods for PLE/PLP to address OHDSI/Atlas#1335. Some notes related to the changes:

- Refactored `EstimationCohortDefinition` and `PredictionCohortDefinition` into **`AnalysisCohortDefinition`**. Also refactored `EstimationConceptSet` and `PredictionCohortDefinition` into **`AnalysisConceptSet`**. This eliminated code redundancy and also enabled me to add methods for importing these portions of the specification in a consistent manner as detailed in the `org.ohdsi.webapi.analysis.converter` and `org.ohdsi.webapi.analysis.importer` namespaces, respectively.
- Updated the StandardizedAnalysisAPI https://github.com/OHDSI/StandardizedAnalysisAPI/pull/17 based on testing that corrected a bug in the way that the Cyclops Prior is specified.
- Removed redundant EstimationTypeEnum from WebAPI in favor of the one in StandardizedAnalysisAPI.

Per discussion with @pavgra, we'll hold this PR until #845 is approved and merged. Once that is done, I'll merge those changes into this branch for review. In the meantime, if there are any concerns with what has been done so far, please let me know and I'll address them while #845 is still in progress.

anthonysena commented:
This is now ready for review. Updates in this commit: https://github.com/OHDSI/WebAPI/pull/861/commits/71a4e27dc7e20f31419e40d446a01ba6f74368ac address OHDSI/Atlas#1221

pavgra commented on src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java:
```diff
@@ -260,8 +262,109 @@ public EstimationAnalysisImpl exportAnalysis(Estimation est) {
+            // Create all of the cohort definitions 
+            // and map the IDs from old -> new
+            Map<Long, Long> cohortIds = new HashMap<>();
+            AnalysisCohortDefinitionImportService cohortImportService = new AnalysisCohortDefinitionImportService(security, userRepository, cohortDefinitionRepository);
```
Why do you use bean as POJO?

pavgra commented on src/main/java/org/ohdsi/webapi/prediction/PredictionServiceImpl.java:
```diff
@@ -187,9 +193,77 @@ public PatientLevelPredictionAnalysisImpl exportAnalysis(int id) {
+            // and map the IDs from old -> new
+            List<BigDecimal> newTargetIds = new ArrayList<>();
+            List<BigDecimal> newOutcomeIds = new ArrayList<>();
+            AnalysisCohortDefinitionImportService cohortImportService = new AnalysisCohortDefinitionImportService(security, userRepository, cohortDefinitionRepository);
```
Why do you use bean as POJO?

pavgra commented on src/main/java/org/ohdsi/webapi/analysis/importer/AnalysisCohortDefinitionImportService.java:
```diff
@@ -0,0 +1,33 @@
+        this.cohortRepository = cohortRepository;
+    }
+    
+    public CohortDefinition persistCohort(final AnalysisCohortDefinition analysisCohortDefinition) {
```
1) Such approach will create duplicates
2) Required method already exists - `DesignImportService.persistCohortOrGetExisting`

anthonysena replied:
Refactored to use `DesignImportService` and added an overloaded method for `persistCohortOrGetExisting` which also checks the cohort name along with the expression hash so that when cohorts are imported, they retain their name.

pavgra commented on src/main/java/org/ohdsi/webapi/analysis/importer/AnalysisConceptSetImportService.java:
```diff
@@ -0,0 +1,31 @@
+    }
+    
+    public ConceptSetDTO persistConceptSet(final AnalysisConceptSet analysisConceptSet) {
+        ConceptSetDTO cs = csConverter.convert(analysisConceptSet);
```
Will create duplicates. I suggest applying the same approach as used for cohorts in `
2\. DesignImportService.persistCohortOrGetExistin`g

anthonysena replied:
So I stopped short of adding in functionality to check the concept set as this feels like it could be fairly complex unless we introduce a hash on the concept set as was done for cohort definition details. Otherwise, I was starting down the path of: identify the concept set by name, then filter by the items. But this requires a number of checks on the various flags for each concept set item, etc. For now, I think we may need to live with duplicate concept sets upon import of a PLE/PLP unless you have some ideas on how to do this without a lot of extra work at this point? This may need to be addressed in 2.8.

anthonysena replied:
The other thought I had here: I could move the code from `org.ohdsi.webapi.analysis.importer.AnalysisConceptSetImportService` and make it part of the `DesignImportService`. This way the design import functions are in a single place and it eliminates the need for a single class in a single namespace (since I was able to remove the `AnalysisCohortDefinitionImportService`)

pavgra replied:
> I could move the code from `org.ohdsi.webapi.analysis.importer.AnalysisConceptSetImportService` and make it part of the `DesignImportService`

Yes, I would do this

pavgra commented on src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java:
```diff
@@ -260,8 +262,109 @@ public EstimationAnalysisImpl exportAnalysis(Estimation est) {
+            // Create all of the concept sets and map
+            // the IDs from old -> new
+            Map<Integer, Integer> conceptSetIdMap = new HashMap<>();
+            AnalysisConceptSetImportService conceptSetImportService = new AnalysisConceptSetImportService(conceptSetService);
```
Why do you use bean as POJO?

pavgra commented on src/main/java/org/ohdsi/webapi/prediction/PredictionServiceImpl.java:
```diff
@@ -187,9 +193,77 @@ public PatientLevelPredictionAnalysisImpl exportAnalysis(int id) {
+            // Create all of the concept sets and map
+            // the IDs from old -> new
+            Map<Integer, Integer> conceptSetIdMap = new HashMap<>();
+            AnalysisConceptSetImportService conceptSetImportService = new AnalysisConceptSetImportService(conceptSetService);
```
Why do you use bean as POJO?

pavgra commented on src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java:
```diff
@@ -75,6 +80,8 @@
             "modifiedBy"
     );
+    
+    private final AnalysisCohortDefinitionToCohortDefinitionConverter cohortConversionService = new AnalysisCohortDefinitionToCohortDefinitionConverter();
```
Why do you instantiate converter manually? Use `conversionService` instead

pavgra commented on src/main/java/org/ohdsi/webapi/common/DesignImportService.java:
```diff
@@ -46,12 +51,21 @@ public CohortDefinition persistCohortOrGetExisting(final CohortDefinition cohort
                 .map(CohortDefinitionDetails::getCohortDefinition);
     }
+    
+    private Optional<CohortDefinition> findCohortByExpressionHashcodeAndName(final CohortDefinitionDetails details, final String cohortName) {
```
I would not copy implementation: I would create `findCohortByExpressionHashcodeAndPredicate` which accepts `CohortDefinitionDetails details` and predicate (lambda) as inputs. The predicate is passed into the `filter`. And two methods override it: `findCohortByExpressionHashcode` which passes `c -> true` as predicate and `findCohortByExpressionHashcodeAndName` which passes `c -> Objects.equals(c.getName(), cohortName)` (use Objects.equals to be null safe)

pavgra commented on src/main/java/org/ohdsi/webapi/prediction/PredictionServiceImpl.java:
```diff
@@ -54,6 +61,8 @@
             "modifiedBy"
     );
+    
+    private final AnalysisCohortDefinitionToCohortDefinitionConverter cohortConversionService = new AnalysisCohortDefinitionToCohortDefinitionConverter();
```
Why do you instantiate converter manually? Use `conversionService` instead

anthonysena commented:
1. Removed `org.ohdsi.webapi.analysis.importer.AnalysisConceptSetImportService` and made it part of the `DesignImportService`
2. Refactored to use `conversionService`
3. Refactored to use `Predicate<CohortDefinition>`

Let me know if you have any other feedback. Thanks!

anthonysena commented:
> Why do you have private final AnalysisCohortDefinitionToCohortDefinitionConverter cohortConversionService = new AnalysisCohortDefinitionToCohortDefinitionConverter();? - what's wrong with converting data via conversionService?

Oversight on my part; pushed a commit that removes this from the code.

> Just for my curiosity, could you explain why do you introduce AnalysisCohortDefinition? - why not utilize existing CohortDTO / CohortDefinition. And what's the reason for separate class for AnalysisConceptSet (doesn't seem that it introduces any new functionality or fields)?

I'm struggling to recall why I did this but I believe I needed to use the `@JsonIgnoreProperties(ignoreUnknown=true)` for these classes to get things working. I can revisit this for 2.8 since I suspect we can drop these extraneous classes or figure out a way to handle the root of the problem without the need for creating more classes. Let me know if you have any concerns with this approach.

pavgra commented on src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java:
```diff
@@ -260,8 +265,109 @@ public EstimationAnalysisImpl exportAnalysis(Estimation est) {
+            // Create the estimation 
+            Estimation est = new Estimation();
+            est.setDescription(analysis.getDescription());
+            est.setName(String.format(ENTITY_COPY_PREFIX, analysis.getName()));
```
Potential collision: I would use the same approach as applied for Pathways (`PathwayController.copy`)

pavgra commented on src/main/java/org/ohdsi/webapi/prediction/PredictionServiceImpl.java:
```diff
@@ -187,9 +199,76 @@ public PatientLevelPredictionAnalysisImpl exportAnalysis(int id) {
+            // Create the prediction analysis
+            PredictionAnalysis pa = new PredictionAnalysis();
+            pa.setDescription(analysis.getDescription());
+            pa.setName(String.format(ENTITY_COPY_PREFIX, analysis.getName()));
```
Potential collision: I would use the same approach as applied for Pathways (`PathwayController.copy`)
